### ISSUE TITLE
Slider - optional field input feature not working

### ISSUE THREAD

**Issue by serenamarie125** [patternfly/patternfly-3#930]
Based on the documentation of the [PatternFly Slider](http://www.patternfly.org/pattern-library/forms-and-controls/slider/) , "The slider can optionally have a field input on the right side to enter the desired value as an alternative to sliding. The slider button should move to indicate the value entered."

When accessing the component on the test page (https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/bootstrap-slider.html?) entering a blue into the input field doesn't affect the slider value.

**Comment by matthewcarleton** [comment]
@serenamarie125 I believe this is b/c of the debate of when JS belongs in pf-core. @jeff-phillips-18 what do you think should we add this for the purpose of the demo?

**Comment by dabeng** [comment]
Hi @serenamarie125 . I reviewed the source code of bootstrap-slider.html page and patternfly.js. It seems that  there are no according code snippets to implement the feature described on the [design doc](http://www.patternfly.org/pattern-library/forms-and-controls/slider/#design).

`
6. Field Input (Optional): The slider can optionally have a field input on the right side to enter the desired value as an alternative to sliding. The slider button should move to indicate the value entered.
`

Maybe it's for this reason that "Slider code example is not available".
![screenshot from 2018-01-27 10-56-51](https://user-images.githubusercontent.com/5689075/35467973-c0ee336e-02db-11e8-8e5e-f8e84461ccdb.png)

**Comment by dabeng** [comment]
@LHinson @cliffpyles @jeff-phillips-18 shall we begin to implement the expected behaviors of slider with input box for now ?

**Comment by dabeng** [comment]
Hi @serenamarie125 @matthewcarleton  I made a **[Demo](https://rawgit.com/dabeng/patternfly/slider-dist/dist/tests/bootstrap-slider.html)** for slider with input box. Please review.

**Comment by matthewcarleton** [comment]
@dabeng it is possible to have the slider update if the value is updated in the input as well?

**Comment by dabeng** [comment]
Hi @matthewcarleton , you need to press "enter" to update slider value.

**Comment by matthewcarleton** [comment]
ah perfect! thanks @dabeng

**Comment by serenamarie125** [comment]
This is looking really good @dabeng !  I have a couple of questions ... 

1. If the user enters a value, and then tabs out of the field, I'd assume the slider would update

1. Can we set valid characters, so it will only accept numbers to be input in the field?

1. Can we perform input validation?  If we implement the valid characters piece, we still need to make sure that the number entered is between the min & max

curious on what @matthewcarleton & @jeff-phillips-18 think here - should #2 and #3 be shown in core examples, or just in JS ?

**Comment by dabeng** [comment]
Great ideas. Thank you very much @serenamarie125 I have improved the **[Demo](https://rawgit.com/dabeng/patternfly/slider-dist/dist/tests/bootstrap-slider.html)** according to your advice:

1. Adds "blur" event handler to response users' tab-out action

2. Adds input validation for text-box adjacent to slider

3. If uses input string, space or invalid numbers beyond given range, warning tooltip will show and auto disappear after 3000ms. At the same time, text-box will revert to its original valid value

4. I did't modify html markup of bootstrap-slider test page and just add the following javascript block at the end of page:
```js
<script>
  $(function () {
    $('[data-provide]+:text').tooltip({ trigger: 'manual' })
      .on('keypress blur', function (e) {
        if (e.which == 13 || e.type === 'blur') {
          var $this = $(this);
          var sd = $this.prev().slider();
          e.preventDefault();
          var max = sd.slider('getAttribute', 'max');
          var min = sd.slider('getAttribute', 'min');
          if (!$.isNumeric(this.value) || (this.value > max || this.value < min)) {
            var warningInfo = $.isNumeric(this.value) ? 'Valid value should be between ' + min
              + ' and ' + max : 'Valid value should be number';
            $this.attr('data-original-title', warningInfo).tooltip('show').addClass('warning');
            this.value = sd.slider('getValue');
            setTimeout(function(){ $this.tooltip('hide'); }, 3000);
          } else {
            if ($this.is('.warning')) {
              $this.tooltip('hide');
            }
            sd.slider('setValue', this.value);
          }
        }
      })
      .prev().each(function () {
        $(this).next().val(this.value);
        $(this).slider().on('slide', function (e) {
          $(e.target).next().val(e.value);
        });
      });
  });
</script>
```

@cliffpyles @mcarrano Any thoughts about slider with text-box ?

**Comment by jeff-phillips-18** [comment]
Seems a bit much to be handling in core but if we have it already no sense it not providing it. I'll let @serenamarie125 and/or @mcarrano decide if the tooltip is necessary or if the field should just use the stander error outline when invalid or prevent invalid chars all together. This is where it get to be too much to provide in the test pages.

**Comment by matthewcarleton** [comment]
I'm with @jeff-phillips-18 here. If we are providing JS we should do it all 👍

**Comment by mcarrano** [comment]
I think that the demo looks great, @dabeng ! Just a couple of comments. I think that a slider should always have some labeling so that the user can see what the min and max values are at either end of the slider bar. Can we add this? It is something that I think was an oversight in the design doc and we can go back and fix that. I do like the tooltip that reflects why the textbox entry was not accepted. Although if you address the first issue, then I think it's less necessary because it should be clear to the user what the min and max values are.

As for whether all of the validation code should be in the core example, I'll let you guys decide. It's true that from a user perspective, when they see an example like this on the website, they expect it to represent behavior as well as appearance. But I also understand the desire to decouple the JS from the core implementations.

**Comment by maryclarke** [comment]
I agree @mcarrano, it would be good to have end points reflected.  Also, what about having a tooltip when the user hovers over the slider indicating what number they are getting ready to change it to?  This may be overkill, but could also help with understanding of the scale.  This would also obviously only be for sliders without detents.

**Comment by dabeng** [comment]
It looks like we don't come to a agreement on design details of slider with text-box. Perhaps we should finish design firstly and then implement according behavior with js code.

**Comment by serenamarie125** [comment]
@jeff-phillips-18 I agree that I'm not sure that the tooltip is the best mechanism, but from a test page perspective, it's nice to see that we are doing some error checking.  I imagine that this would be used in a form, and would then have consistent error handling for all the form elements.

**Comment by serenamarie125** [comment]
@dabeng great work, if there's a way to limit the valid characters for the text box, it would be nice to see if you could not accept alpha characters, meaning if the user types them, it wouldn't show up in the text field.  Again, this may be more of a nice to have, but bringing it up in case it's easy to fix!

**Comment by mcarrano** [comment]
@dabeng The tooltip the way you have it is fine to report an entry error in this demo. The only thing we ask whether it's possible to just constrain the text box to only accept a numeric (whole number or decimal) value. If not, we can keep as is. I'd also ask that you add end points (i.e. 0 and 100) to your example to provide more feedback on the valid range. We can update the design documentation later, if needed.

**Comment by dabeng** [comment]
Hi @serenamarie125 @mcarrano .  Do you mean that we could remove tooltip ?

**Comment by dabeng** [comment]
Additionally, adding end points and character validation are easy to implement. Trust me.

**Comment by mcarrano** [comment]
@dabeng I don't have a problem with leaving the tooltip in. Do you, @serenamarie125 ?

**Comment by serenamarie125** [comment]
@dabeng that's fine, you can leave it.  Thanks!

**Comment by serenamarie125** [comment]
looks great, thanks @dabeng are you putting up a PR for this?

**Comment by cliffpyles** [comment]
This looks great @dabeng! Awesome work!

**Comment by LHinson** [comment]
Closing this issue since it is resolved by #965
